bindings_ffi/src/v2.rs

```diff
@@ -279,46 +283,90 @@ impl From<FfiV2SubscribeRequest> for SubscribeRequest {
     }
 }
 
-#[derive(uniffi::Object)]
-pub struct FfiV2Subscription {
-    inner_subscription:
-        Arc<futures::lock::Mutex<xmtp_api_grpc::grpc_api_helper::GrpcMutableSubscription>>,
+#[uniffi::export(callback_interface)]
+pub trait FfiV2SubscriptionCallback: Send + Sync {
+    fn on_message(&self, message: FfiEnvelope);
 }
 
-impl From<xmtp_api_grpc::grpc_api_helper::GrpcMutableSubscription> for FfiV2Subscription {
-    fn from(subscription: xmtp_api_grpc::grpc_api_helper::GrpcMutableSubscription) -> Self {
-        Self {
-            inner_subscription: Arc::new(futures::lock::Mutex::new(subscription)),
-        }
-    }
+/// Subscription to a stream of V2 Messages
+#[derive(uniffi::Object)]
+pub struct FfiV2Subscription {
+    tx: mpsc::Sender<FfiV2SubscribeRequest>,
+    abort: AbortHandle,
+    // we require Arc<Mutex<>> here because uniffi doesn't like &mut or self in impl
+    #[allow(clippy::type_complexity)]
+    handle: Arc<Mutex<Option<JoinHandle<()>>>>,
 }
 
 #[uniffi::export(async_runtime = "tokio")]
 impl FfiV2Subscription {
-    pub async fn next(&self) -> Result<FfiEnvelope, GenericError> {
-        let mut sub = self.inner_subscription.lock().await;
-
-        let result = sub.next().await;
-        match result {
-            Some(Ok(envelope)) => Ok(envelope.into()),
-            Some(Err(err)) => Err(GenericError::Generic {
-                err: err.to_string(),
-            }),
-            None => Err(GenericError::Generic {
-                err: "stream closed".to_string(),
-            }),
+    /// End the subscription, waiting for the subscription to close entirely.
+    /// # Errors
+    /// * Errors if subscription event task encounters join error
+    pub async fn end(&self) -> Result<(), GenericError> {
+        if self.abort.is_finished() {
+            return Ok(());
         }
+
+        let mut handle = self.handle.lock().await;
+        let handle = handle.take();
+        if let Some(h) = handle {
+            h.abort();
+            h.await.map_err(|_| GenericError::Generic {
+                err: "subscription event loop join error".into(),
+            })?;
+        }
+        Ok(())
     }
 
+    /// Check if the subscription is closed
+    pub fn is_closed(&self) -> bool {
+        self.abort.is_finished()
+    }
+
+    /// Update subscription with new topics
     pub async fn update(&self, req: FfiV2SubscribeRequest) -> Result<(), GenericError> {
-        let mut sub = self.inner_subscription.lock().await;
-        sub.update(req.into()).await?;
+        self.tx.send(req).await.map_err(|_| GenericError::Generic {
```

Comment:
What is the pro/con of sending it through via mpsc, compared to calling it on the subscription directly like in the previous implementation?

Reply:
That's a good question, other than the usual edge-casey footguns with mpsc,

I think the biggest downside is the extra memory overhead for the channel itself, which in this case is bounded to 10 updates. Since it is bounded, if `update` is called continually like in a loop we might lose some updates after reaching the limit on the channel. This shouldn't really be an issue, though, since the caller should be putting the topics in a `Vec<>` to the channel `FfiV2SubscribeRequest` type anyway. The limit is needed to avoid a infinitely growing channel.

Reply:
That makes sense -  I guess more specifically, what would be the drawback of storing the subscription directly on FFIV2Subscription, and directly updating it here as in the old implementation, versus the new approach of storing the tx and sending a message over it?

Reply:
By storing the subscription locally in the task, we can avoid using a `Mutex` to access the subscription. This has a few benefits, and it more generally/lightly follows the `actor` model of concurrency rather than locking:

- It's easier to reason about what's happening to the subscription, because all the events are handled in one event loop the task is spawning
- we avoid any footguns we might run into with `Mutex's`, which lets us avoid a class of errors that occurs with `Mutexes`, which makes `update` more difficult to misuse outside of libxmtp. We still use `Mutexes` a bunch which is fine, esp considering this is with uniffi, but more mutexes means more potential for deadlocks and the such. Mutexes are more performant, though, but I doubt using a channel vs mutex will make a huge diff in that direction here
- if we want to modify the subscription in other ways, can extend it by just adding new events to the loop

I guess its a bit arbitrary, when writing streams I do default to this strategy rather than `Mutex`'s, unless absolutely necessary for performance or to satisfy the borrow checker w.r.t uniffi or the like